From 9aea08755a8e01e176fb3a0b5ad92e663473527b Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Tue, 1 Feb 2022 20:59:18 +0100
Subject: [PATCH] sample: ipc : add service bind unbind support in openamprsc
 table sample

Allow to bin a rpmsg_tty channel and unbind it

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
---
 .../ipc/openamp_rsc_table/src/main_remote.c   | 30 +++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c b/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
index 093b14a41772..cbf79099eecd 100644
--- a/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
+++ b/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c
@@ -121,9 +121,39 @@ static void receive_message(unsigned char **msg, unsigned int *len)
 	}
 }
 
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ept)
+{
+	LOG_INF("destroy end point name %s\n", log_strdup(ept->name));
+	/*
+	 * FIXME: need to set endpoint name to void to not send a ns destroy
+	 * announcement that generates an error on Linux side
+	 */
+	ept->name[0] = 0;
+	rpmsg_destroy_ept(ept);
+}
+
 static void new_service_cb(struct rpmsg_device *rdev, const char *name,
 			   uint32_t src)
 {
+	int ret;
+
+	if (strcmp(name, "rpmsg-tty") == 0  && !tty_ept[1].rdev) {
+		tty_ept[1].priv = &tty_msg[1];
+		ret = rpmsg_create_ept(&tty_ept[1], rpdev, "rpmsg-tty",
+				       RPMSG_ADDR_ANY, src,
+				       rpmsg_recv_tty_callback,
+				       rpmsg_service_unbind);
+		printk("APO: %s (%d) : ept address %x, %x\r\n", __func__, __LINE__,
+		       tty_ept[1].addr, tty_ept[1].dest_addr);
+		if (ret != 0)
+			LOG_ERR("Creating remote endpoint %s"
+				" failed wirh error %d", log_strdup(name), ret);
+		else {
+			LOG_INF("request to bind new service %s\n", log_strdup(name));
+			rpmsg_send(&tty_ept[1], "bound", sizeof("bound"));
+		}
+		return;
+	}
 	LOG_ERR("%s: unexpected ns service receive for name %s\n",
 		__func__, name);
 }
